robot_sort/robot_sort.py

```diff
@@ -81,32 +81,72 @@ def set_light_on(self):
         Turn on the robot's light
         """
         self._light = "ON"
+
     def set_light_off(self):
         """
         Turn off the robot's light
         """
         self._light = "OFF"
+
     def light_is_on(self):
         """
         Returns True if the robot's light is on and False otherwise.
         """
         return self._light == "ON"
 
+    def side_step(self):
+        self.swap_item()
+        self.move_left()
+        self.swap_item()
+        self.move_right()
+
     def sort(self):
         """
         Sort the robot's list.
         """
+        # Compare the held item with the item in front of the robot:
+        # If the held item's value is greater, return 1.
+        # If the held item's value is less, return -1.
+        # If the held item's value is equal, return 0.
+        # If either item is None, return None.
         # Fill this out
-        pass
+
+        # def bubble_sort(arr):
+        #     for i in range(0, len(arr) - 1):
+        #         for j in range(i, len(arr)):
+        #             if arr[i] > arr[j]:
+        #                 arr[j], arr[i] = arr[i], arr[j]
+        # bubble_sort(self._list)
+
+        # Now, with robot methods!
+
+        while not self.light_is_on():  # exit condition
+            self.set_light_on()
+            while self.can_move_right():  # will loop until end of list
+                self.swap_item()
+                self.move_right()
+                if self.compare_item() == 1:  # array swapping!
+                    # side_step swaps adjacent elements
+                    self.side_step()
+                    self.set_light_off()  # exit the loop
+                else:
+                    # side_step except doesn't swap the elements
+                    self.move_left()
+                    self.swap_item()
+                    self.move_right()
+            if not self.light_is_on():
+                while self.can_move_left():
+                    self.move_left()
```

Comment:
Good job with this. An alternative approach would be selection sort which could be implemented like this:

self.swap_item()
self.set_light_on()

while self.light_is_on():
(move right and find the smallest item that has not been sorted)
____while self.move_right():
________if self.compare_item() > 0:
____________self.swap_item()
( if the robot is at the end of the list and all items are sorted, we're done)
________if self.compare_item() is None and self.can_move_right() is False:
(place the current item down in the list and turn light off to indicate the list is sorted)
____________self.swap_item()
____________self.set_light_off()
____________break
________else:
( find the spot in the list that is holding None)
( place the item in that position)
(move right and swap)
____________while self.move_left():
________________if self.compare_item() is None:
________________self.swap_item()
________________self.move_right()
________________self.swap_item()
________________break